File: Kalorie/Kalorie/Core/UseCases/LinkOrMergeCredentialUseCase.swift
//
//  LinkOrMergeCredentialUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 07.08.2026.
//

import FirebaseAuth
import Foundation

enum LinkOrMergeCredentialError: Error {
    case accountExistsWithAnotherProvider
}

protocol LinkOrMergeCredentialUseCaseProtocol {
    func callAsFunction(credential: AuthCredential) async throws
}

struct LinkOrMergeCredentialUseCase: LinkOrMergeCredentialUseCaseProtocol {

    // MARK: - Properties

    private let authProvider: any AuthProviderProtocol
    private let authCommandProvider: any AuthCommandProviderProtocol
    private let migrateAnonymousData: any MigrateAnonymousDataUseCaseProtocol

    // MARK: - Init

    init(
        authProvider: any AuthProviderProtocol,
        authCommandProvider: any AuthCommandProviderProtocol,
        migrateAnonymousData: any MigrateAnonymousDataUseCaseProtocol
    ) {
        self.authProvider = authProvider
        self.authCommandProvider = authCommandProvider
        self.migrateAnonymousData = migrateAnonymousData
    }

    // MARK: - Functions

    func callAsFunction(credential: AuthCredential) async throws {
        let anonymousUserId = authProvider.userId
        do {
            try await authCommandProvider.link(with: credential)
        } catch {
            guard let nsError = error as NSError? else { throw error }

            if nsError.code == AuthErrorCode.emailAlreadyInUse.rawValue {
                throw LinkOrMergeCredentialError.accountExistsWithAnotherProvider
            }

            guard
                let anonymousUserId,
                nsError.code == AuthErrorCode.credentialAlreadyInUse.rawValue
            else { throw error }

            try await migrateAnonymousData.migrate(fromAnonymousUserId: anonymousUserId, credential: credential)
        }
    }
}

#if DEBUG
struct LinkOrMergeCredentialUseCaseFake: LinkOrMergeCredentialUseCaseProtocol {

    // MARK: - Properties

    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction(credential: AuthCredential) async throws {
        if shouldThrow { throw URLError(.unknown) }
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/LinkOrMergeCredentialUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthCommandProviderProtocol
import antoni.kalorie.core.auth.AuthProviderProtocol
import com.google.firebase.auth.AuthCredential
import com.google.firebase.auth.FirebaseAuthUserCollisionException
import kotlinx.coroutines.CancellationException

sealed class LinkOrMergeCredentialError : Exception() {
    data object AccountExistsWithAnotherProvider : LinkOrMergeCredentialError()
}

interface LinkOrMergeCredentialUseCaseProtocol {
    suspend operator fun invoke(credential: AuthCredential)
}

class LinkOrMergeCredentialUseCase(
    private val authProvider: AuthProviderProtocol,
    private val authCommandProvider: AuthCommandProviderProtocol,
    private val migrateAnonymousData: MigrateAnonymousDataUseCaseProtocol,
) : LinkOrMergeCredentialUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(credential: AuthCredential) {
        val anonymousUserId = authProvider.userId
        try {
            authCommandProvider.link(credential)
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            val errorCode = (error as? FirebaseAuthUserCollisionException)?.errorCode
            if (errorCode == EMAIL_ALREADY_IN_USE) throw LinkOrMergeCredentialError.AccountExistsWithAnotherProvider
            if (anonymousUserId == null || errorCode != CREDENTIAL_ALREADY_IN_USE) throw error
            migrateAnonymousData.migrate(anonymousUserId, credential)
        }
    }

    private companion object {
        const val EMAIL_ALREADY_IN_USE = "ERROR_EMAIL_ALREADY_IN_USE"
        const val CREDENTIAL_ALREADY_IN_USE = "ERROR_CREDENTIAL_ALREADY_IN_USE"
    }
}
